### PR TITLE
unified property access

### PR DESCRIPTION
**Pull request by devisnik**
- handles all the property access via the extension 
- reduces artifact method parameters by handing the project via constructor
- gets rid of Artifacts interface instead relying on Groovy's runtime method resolution

Note: This is completely untested. Thus it should not be merged, but might be useful for further discussion. :x:

### CODE REVIEW

File: core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy [hunk 1]
```diff
@@ -1,34 +1,61 @@
...
+        prop('dryRun')
+    }
+
+    def getPublishVersion() {
+        prop('publishVersion') ?: version
+    }
+
+    private String prop(String propertyName) {
+        project.hasProperty(propertyName) ? project.property(propertyName) : getProperty(propertyName)
+    }
```
Comment (devisnik) [inline core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy]: I think having the property resolution here along with the definitions makes it a bit more readable.

File: core/src/main/groovy/com/novoda/gradle/release/JavaArtifacts.groovy [hunk 2]
```diff
@@ -1,29 +1,34 @@
...
 import org.gradle.api.Project
 import org.gradle.api.tasks.bundling.Jar
 
-class JavaArtifacts implements Artifacts {
+class JavaArtifacts {
+
+    private Project project
+
+    JavaArtifacts(Project project) {
+        this.project = project
```
Comment (devisnik) [inline core/src/main/groovy/com/novoda/gradle/release/JavaArtifacts.groovy]: Holds an instance of the plugin object to reduce the method parameters.

File: core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy [hunk 3]
```diff
@@ -1,34 +1,61 @@
...
+    def getBintrayKey() {
+        prop('bintrayKey')
+    }
+
+    def getDryRun() {
+        prop('dryRun')
+    }
+
+    def getPublishVersion() {
+        prop('publishVersion') ?: version
```
Comment (ouchadam) [inline core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy]: this has lost the deprecated `version` logic

File: core/src/main/groovy/com/novoda/gradle/release/AndroidArtifacts.groovy [hunk 4]
```diff
@@ -1,41 +1,48 @@
...
         }
     }
 
-    def javadocJar(String publicationName, Project project) {
-        def androidJavadocs = project.task(publicationName + 'AndroidJavadocs', type: Javadoc) {
+    def javadocJar(String publicationName) {
+        Task androidJavadocs = project.task(publicationName + 'AndroidJavadocs', type: Javadoc) {
             source = project.android.sourceSets.main.java.srcDirs
-            classpath += project.files(project.android.getBootClasspath().join(File.pathSeparator))
+            classpath += project.android.bootClasspath
```
Comment (ouchadam) [inline core/src/main/groovy/com/novoda/gradle/release/AndroidArtifacts.groovy]: this still works? nice

File: core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy (hunk 3 above)
Comment (devisnik) [inline core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy]: Is that changing the behaviour? Thought it would work as before.

File: core/src/main/groovy/com/novoda/gradle/release/AndroidArtifacts.groovy (hunk 4 above)
Comment (devisnik) [inline core/src/main/groovy/com/novoda/gradle/release/AndroidArtifacts.groovy]: I hope it does. ;)
I couldn't really test it. Would I have to deploy to a local repo for that?
Can't you just link it to some sample similar to what we do in the command-plugin or the sqlite-analyzer?

File: core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy (hunk 1 above)
Comment (devisnik) [inline core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy]: But it does not work, StackOverflow, so :x: :x: 
Seems that `getProperty(propertyName)` leads to another call of the getter. 

I'll try to fix that, it's a pain with the current setup though.

Comment (ouchadam) [inline core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy]: get property is meant to delegate to the extension?

Comment (devisnik) [inline core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy]: We are inside the extension here. So this is the regular Groovy property resolution mechanism. 
I've fixed the StackOverflow now, accessing the property directly.

File: core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy (hunk 3 above)
Comment (ouchadam) [inline core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy]: my bad, I thought the prop didn't use the extension value. :+1:

**Comment by devisnik** [comment]
I now tested this deploying to mavenLocal and then switching to that version. Quite painful, especially when trying to fix something. :/

File: core/src/main/groovy/com/novoda/gradle/release/AndroidArtifacts.groovy (hunk 4 above)
Comment (blundell) [inline core/src/main/groovy/com/novoda/gradle/release/AndroidArtifacts.groovy]: lol, lets not merge untested code ... reply back when 100% sure of this

File: core/src/main/groovy/com/novoda/gradle/release/AndroidArtifacts.groovy [hunk 5]
```diff
@@ -1,41 +1,48 @@
...
             classpath += project.android.libraryVariants.toList().first().javaCompile.classpath
             classpath += project.android.libraryVariants.toList().first().javaCompile.outputs.files
         }
 
         project.task(publicationName + 'AndroidJavadocsJar', type: Jar, dependsOn: androidJavadocs) {
             classifier = 'javadoc'
-            from androidJavadocs.destinationDir
+            from androidJavadocs
```
Comment (blundell) [inline core/src/main/groovy/com/novoda/gradle/release/AndroidArtifacts.groovy]: can you explain this difference?

File: core/src/main/groovy/com/novoda/gradle/release/JavaArtifacts.groovy [hunk 6]
```diff
@@ -1,29 +1,34 @@
...
             classifier = 'sources'
             from project.sourceSets.main.allSource
         }
     }
 
-    def javadocJar(String publicationName, Project project) {
+    def javadocJar(String publicationName) {
         project.task(publicationName + 'JavadocJar', type: Jar) {
             classifier = 'javadoc'
             from project.javadoc.destinationDir
```
Comment (blundell) [inline core/src/main/groovy/com/novoda/gradle/release/JavaArtifacts.groovy]: so this `from` method is not the method you renamed below on line 26?

File: core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy [hunk 7]
```diff
@@ -1,11 +1,13 @@
...
 
+import org.gradle.api.Project
+
 class PublishExtension {
 
     String repoName = 'maven'
     String userOrg
-
     String groupId
+
```
Comment (blundell) [inline core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy]: why this white line change? Now the grouping seems off

File: core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy [hunk 8]
```diff
@@ -14,21 +16,33 @@ class PublishExtension {
...
     String[] licences = ['Apache-2.0']
 
     String uploadName = ''
+
     String description
     String website = ''
     String issueTracker = ''
     String repository = ''
     boolean autoPublish = true
+    String bintrayUser
```
Comment (blundell) [inline core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy]: why did you move this?

File: core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy [hunk 9]
```diff
@@ -14,21 +16,33 @@ class PublishExtension {
...
     String[] publications = ['maven']
 
+    private Project project
+
+    PublishExtension(Project project) {
+        this.project = project
+    }
+
+    @Override
+    Object getProperty(final String property) {
```
Comment (blundell) [inline core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy]: why is this final

File: core/src/main/groovy/com/novoda/gradle/release/ReleasePlugin.groovy [hunk 10]
```diff
@@ -8,36 +8,31 @@ import org.gradle.api.publish.maven.MavenPublication
...
-
-                    artifacts.all(it.name, project).each {
-                        delegate.artifact it
-                    }
-
-                    from artifacts.from(project)
+                    groupId extension.groupId
+                    artifactId extension.artifactId
+                    version extension.publishVersion ?: extension.version
+                    artifacts = projectArtifacts.all(it.name)
```
Comment (blundell) [inline core/src/main/groovy/com/novoda/gradle/release/ReleasePlugin.groovy]: artifacts uses the field when all others use the setters?

File: core/src/main/groovy/com/novoda/gradle/release/AndroidArtifacts.groovy (hunk 4 above)
Comment (devisnik) [inline core/src/main/groovy/com/novoda/gradle/release/AndroidArtifacts.groovy]: This PR is only a suggestion, it's not meant to be merged.
I tested now locally by deploying to the local maven dir and using that version to re-deploy (after increasing the version).  Seems to work, would need double-checking though.

If we agree that the changes outlined here make sense, I'd be happy to bring this PR into a mergeable state.

File: core/src/main/groovy/com/novoda/gradle/release/AndroidArtifacts.groovy (hunk 5 above)
Comment (devisnik) [inline core/src/main/groovy/com/novoda/gradle/release/AndroidArtifacts.groovy]: We only depend on the outputs of the `androidJavadocs` now, which currently happens to be `destinationDir`.

File: core/src/main/groovy/com/novoda/gradle/release/JavaArtifacts.groovy (hunk 6 above)
Comment (devisnik) [inline core/src/main/groovy/com/novoda/gradle/release/JavaArtifacts.groovy]: No, it's not. I was confused as well.

File: core/src/main/groovy/com/novoda/gradle/release/ReleasePlugin.groovy (hunk 10 above)
Comment (devisnik) [inline core/src/main/groovy/com/novoda/gradle/release/ReleasePlugin.groovy]: It uses the setter as well.
From the javaDoc of `MavenPublication`:
```
     * publishing {
     *   publications {
     *     maven(MavenPublication) {
     *       from components.java
     *       artifacts = ["my-custom-jar.jar", sourceJar]
     *     }
     *   }
     * }
```

Does not work without the `=`. Not sure, why. Might be that the getter is returning `MavenArtifactSet` whereas the setter expects just an `Iterable`.

File: core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy (hunk 7 above)
Comment (devisnik) [inline core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy]: Yep, should be reverted.

File: core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy (hunk 9 above)
Comment (devisnik) [inline core/src/main/groovy/com/novoda/gradle/release/PublishExtension.groovy]: Generated by AS. Should align with the general policy. 
Might move the `private` up to `project`. ;)

**Comment by devisnik** [comment]
I close this.
Seems the property resolution inside the plugin extension does not work as expected. Need to look into that. :/
